### PR TITLE
Refactoring TransportSimulateBulkAction to not extend TransportBulkAction

### PR DESCRIPTION
**Pull request by masseyke**
This refactors TransportSimulateBulkAction and TransportBulkAction to inherit from a new common abstract class, rather than having TransportSimulateBulkAction extend TransportBulkAction. No user-observable functionality has changed. The simulate code no longer runs the ShardBulkAction logic (which is not needed for simulate anyway).
The motivation for this change is so that we remove the risk that a simulate request changes any state in the system (index data, cluster state, etc). The tradeoff is that the simulate path now executes less of the logic used by a bulk request (however, most of that logic is irrelevant to a simulation anyway). The core logic of traversing relevant indices and executing pipelines remains shared between simulate and normal bulk.

### CODE REVIEW

**Comment by masseyke** [comment]
Note: This code was almost entirely taken from the work of @henningandersen at https://github.com/elastic/elasticsearch/compare/main...henningandersen:elasticsearch:improved_simulate_action?expand=1#diff-97ba6f4f3881131b999ea982aa5ab22b31bad9eaf7aa1d14beda660309ae60b7R101

**Comment by masseyke** [comment]
The build is currently broken because https://auth.docker.io/ is having problems.

**Review by parkertimmins — APPROVED** [review]
This looks good to me, nice refactor! That said, I'd definitely suggest not merging until a someone with a better understanding of this code has taken a look

File: server/src/main/java/org/elasticsearch/action/bulk/TransportAbstractBulkAction.java [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+                            executor.execute(runnable);
+                        }
+                    }
+                }
+            },
+            executor
+        );
+    }
+
+    protected abstract boolean shouldStoreFailure(String indexName, Metadata metadata, long epochMillis);
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportAbstractBulkAction.java:274]: Can you add javadocs for this method?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 2]
```diff
@@ -781,13 +552,18 @@ public boolean isForceExecution() {
      * @return true if the given index name corresponds to a data stream with a failure store,
      * or if it matches a template that has a data stream failure store enabled.
      */
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:554]: This is probably the javadoc you can copy over

File: server/src/main/java/org/elasticsearch/action/bulk/TransportSimulateBulkAction.java [hunk 3]
```diff
@@ -28,60 +25,53 @@
...
-        Set<String> failureStoresToBeRolledOver,
-        long startTime
+        long relativeStartTime
     ) {
         final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
         for (int i = 0; i < bulkRequest.requests.size(); i++) {
-            DocWriteRequest<?> request = bulkRequest.requests.get(i);
-            assert request instanceof IndexRequest; // This action is only ever called with IndexRequests
+            DocWriteRequest<?> docRequest = bulkRequest.requests.get(i);
+            assert docRequest instanceof IndexRequest; // This action is only ever called with IndexRequests
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportSimulateBulkAction.java:72]: Can you change the comment into an assertion message?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportSimulateBulkAction.java [hunk 4]
```diff
@@ -111,4 +103,9 @@ protected IngestService getIngestService(BulkRequest request) {
         IngestService rawIngestService = super.getIngestService(request);
         return new SimulateIngestService(rawIngestService, request);
     }
+
+    @Override
+    protected boolean shouldStoreFailure(String indexName, Metadata metadata, long time) {
+        return false;
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportSimulateBulkAction.java:109]: Can you add a comment about why this is always false?

**Review by dakrone — APPROVED** [review]
LGTM, I left some comments, but they're all around comment stuff. Thanks Keith!

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**